components/automate-gateway/pkg/limiter/limiter.go

```diff
@@ -0,0 +1,88 @@
+package limiter
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+)
+
+var ErrLimitExceeded = errors.New("Resource limit exceeded")
+
+type Limiter interface {
+	TakeTicket() (Ticket, error)
+}
+
+type Ticket interface {
+	Return()
+}
+
+type inflightRequestLimiter struct {
+	lock        sync.Mutex
+	maxRequests int
+	inflight    int
+	name        string
+}
+
+type inflightRequestLimiterTicket struct {
+	limiter  *inflightRequestLimiter
+	returned bool
+}
+
+func NewInflightRequestLimiter(name string, maxRequests int) Limiter {
+	return &inflightRequestLimiter{
+		name:        name,
+		maxRequests: maxRequests,
+	}
+}
+
+func (l *inflightRequestLimiter) TakeTicket() (Ticket, error) {
+	taken := false
+	cur := 0
+	max := 0
+	l.lock.Lock()
+	cur = l.inflight
+	max = l.maxRequests
+	if cur < max {
+		l.inflight++
+		taken = true
+	}
+	l.lock.Unlock()
+
+	if !taken {
+		err := fmt.Errorf("resource=%s cur=%d max=%d: %w",
```

Comment:
Now I'm confused. Is `errors.Is(err, ErrLimitExceeded)` _true_? But I'm also tired. Will have another look in the morning.

Reply:
the tests below say yes?

Reply:
I believe it will be. The `%w` formatter does more than just take the error's message and keeps track of the wrapped errors. errors.Is will return true even for deeply nested versions of this error.

Reply:
Yeah, I hadn't claimed it wasn't true, I just didn't see how. Thanks @stevendanna, that's indeed the case:

> If the format specifier includes a %w verb with an error operand, **the returned error will implement an Unwrap method returning the operand.** It is invalid to include more than one %w verb or to supply it with an operand that does not implement the error interface. The %w verb is otherwise a synonym for %v.

https://golang.org/pkg/fmt/#Errorf

Nice. We could look into automatically refactoring all our error wraps...